File: Google/Golang/lowest-common-ancestor-of-a-binary-tree.go
/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *ListNode
 *     Right *ListNode
 * }
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    res := root
    helper(root, p, q, &res)
    return res
}

func helper(root, p, q *TreeNode, res **TreeNode) bool {
    if root == nil {
        return false
    }
    flag := root == p || root == q
    left := helper(root.Left, p, q, res)
    right := helper(root.Right, p, q, res)
    if (left && right) || (flag && left) || (flag && right) {
        *res = root
    }
    return left || right || flag
}

File: Tree/C++/lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        TreeNode *ans;
        helper(root, p->val, q->val, ans);
        return ans;
    }
    
    bool helper(TreeNode *root, int &pVal, int &qVal, TreeNode *&ans) {
        bool left, curr, right;
        if (root == NULL)
            return false;
        left = helper(root->left, pVal, qVal, ans);
        curr = ((root->val == pVal) || (root->val == qVal));
        right = helper(root->right, pVal, qVal, ans);
        if ((left && curr) || (curr && right) || (left && right))
            ans = root;
        return left || curr || right;
    }
};
